services/venice-controller/src/main/java/com/linkedin/venice/controller/systemstore/SystemStoreHealthCheckService.java

```diff
@@ -0,0 +1,179 @@
+  private final MetaStoreReader metaStoreReader;
+  private final MetaStoreWriter metaStoreWriter;
+
+  private final int checkPeriodInSeconds;
+  private final AtomicBoolean isRunning = new AtomicBoolean(false);
+  private final AtomicReference<Set<String>> unhealthySystemStoreSet = new AtomicReference<>();
+
+  private ScheduledExecutorService checkServiceExecutor;
+
+  public SystemStoreHealthCheckService(
+      ReadWriteStoreRepository storeRepository,
+      MetaStoreReader metaStoreReader,
+      MetaStoreWriter metaStoreWriter,
+      PushStatusStoreReader pushStatusStoreReader,
+      PushStatusStoreWriter pushStatusStoreWriter,
+      int systemStoreCheckPeriodInSeconds) {
+    this.storeRepository = storeRepository;
+    this.metaStoreWriter = metaStoreWriter;
+    this.metaStoreReader = metaStoreReader;
+    this.pushStatusStoreWriter = pushStatusStoreWriter;
+    this.pushStatusStoreReader = pushStatusStoreReader;
+    this.checkPeriodInSeconds = systemStoreCheckPeriodInSeconds;
+    this.unhealthySystemStoreSet.set(new HashSet<>());
+  }
+
+  /**
+   * Return unhealthy system store name set. This API is expected to be called by parent controller.
+   */
+  public Set<String> getUnhealthySystemStoreSet() {
+    return unhealthySystemStoreSet.get();
+  }
+
+  @Override
+  public boolean startInner() {
+    checkServiceExecutor = Executors.newScheduledThreadPool(1);
+    isRunning.set(true);
+    checkServiceExecutor.scheduleWithFixedDelay(
+        new SystemStoreHealthCheckTask(),
+        checkPeriodInSeconds,
+        checkPeriodInSeconds,
+        TimeUnit.SECONDS);
+    LOGGER.info("System store health check executor service is started.");
+    return true;
+  }
+
+  @Override
+  public void stopInner() {
+    isRunning.set(false);
+    checkServiceExecutor.shutdown();
+    try {
+      if (!checkServiceExecutor.awaitTermination(5, TimeUnit.SECONDS)) {
+        LOGGER.warn(
+            "System store health check executor service is not terminated after 5 seconds, will be force shutdown now.");
+        checkServiceExecutor.shutdownNow();
+      }
+    } catch (InterruptedException e) {
+      currentThread().interrupt();
+    }
+    LOGGER.info("System store health check executor service is shutdown.");
+  }
+
+  private class SystemStoreHealthCheckTask implements Runnable {
+    @Override
+    public void run() {
+      if (!isRunning.get()) {
+        return;
+      }
+
+      Set<String> newUnhealthySystemStoreSet = new HashSet<>();
+      Map<String, Long> systemStoreToHeartbeatTimestampMap = new VeniceConcurrentHashMap<>();
+      for (Store store: storeRepository.getAllStores()) {
+        if (!isRunning.get()) {
+          return;
+        }
+        if (!VeniceSystemStoreUtils.isUserSystemStore(store.getName())) {
+          continue;
+        }
+        // It is also possible that the store is a new store and is being empty pushed.
+        if (store.getCurrentVersion() == 0) {
```

Comment:
So there are several error scenarios:
1. Store doesn't have system store.
2. Store has system store without a current version.
3. Store has an active current version, but it is not ingesting.

I see 1) is missing, and is it expected?

Reply:
It is expected. System store is always be created as part of store creation. It is not an independent store object but it is inside user store's ZK node. I don't think we need to check it here. 
The truly important action here should be (2) which is empty push failure (3) it goes back silently

Reply:
What if the system store creation is problematic? For example, empty push didn't hit child controller at all, and sys stores won't be auto created.

Reply:
>     
    // Get all the regular Venice stores
    List<Store> regularVeniceStores = regularStoreRepository.getAllStores();
    List<Store> allStores = new ArrayList<>(regularVeniceStores);
    // So far, only consider meta system store.
    Store zkSharedStoreForMetaSystemStore =
        systemStoreRepository.getStore(VeniceSystemStoreType.META_STORE.getZkSharedStoreName());
    Store zkSharedStoreForDaVinciPushStatusSystemStore =
        systemStoreRepository.getStore(VeniceSystemStoreType.DAVINCI_PUSH_STATUS_STORE.getZkSharedStoreName());
    // Populate all the systems stores for the regular Venice stores if they are enabled.
    regularVeniceStores.forEach(store -> {
      if (store.isStoreMetaSystemStoreEnabled()) {
        allStores.add(new SystemStore(zkSharedStoreForMetaSystemStore, VeniceSystemStoreType.META_STORE, store));
      }
      if (store.isDaVinciPushStatusStoreEnabled()) {
        allStores.add(
            new SystemStore(
                zkSharedStoreForDaVinciPushStatusSystemStore,
                VeniceSystemStoreType.DAVINCI_PUSH_STATUS_STORE,
                store));
      }
    });
    return allStores;


This is the impl of getAllStores() in store repo. My understanding is for a regular store, two system stores always exist as the flag will be true for them. 

I see addVersionOnly tries to update the store to have two system stores. Are you suggesting that this method might not be hit for some region?

The reason that I did not check it in the current implementation is if I am going to check it I might need to iterate the store list twice. (Although it is affordable).

Reply:
Let me resolve all other issues first. I think I am 50/50 towards adding this check or not, if you think this could actually happen, I think I can add it, probably tweak the structure a bit to make unit test coverage happy.

Reply:
Synced offline, I've updated the PR to reflected the changes.